crates/libs/core/src/mem.rs

```diff
@@ -0,0 +1,68 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+/// A pool to hold boxes to extend their lifetime.
+/// This is typically useful to build raw COM structures that require
+/// raw pointers in struct fields.
+#[derive(Debug, Default)]
+pub struct BoxPool {
+    inner: Vec<Box<dyn 'static + std::any::Any>>,
+}
+
+impl BoxPool {
+    pub fn new() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    /// Push a box into the pool, and return a raw pointer to the boxed value.
+    /// The pointer is valid as long as the pool is alive.
+    pub fn push<T: 'static>(&mut self, b: Box<T>) -> *const T {
+        let raw = b.as_ref() as *const T;
+        self.inner.push(b);
+        raw
+    }
+
+    /// Push a Vec<T> into the pool, and return its length and a raw pointer to its data.
+    /// The pointer is valid as long as the pool is alive.
+    pub fn push_vec<T: 'static>(&mut self, v: Vec<T>) -> (usize, *const T) {
+        let len = v.len();
+        let raw = v.as_ptr();
+        // Convert Vec<T> to Box<dyn Any>.
+        let boxed_v = Box::new(v);
```

Comment:
Does this involve extra memory allocation or is it complete transparent?

Reply:
This requires allocation. I tried with Box<[T]> there is problem with Sized and [T], push a boxed vector is good enough for now.